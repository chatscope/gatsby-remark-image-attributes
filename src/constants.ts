export const CSS_CLASS = 'gatsby-img-attributes';
