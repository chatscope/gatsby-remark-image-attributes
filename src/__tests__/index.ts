import { PluginResult, Options } from '../index.d';

import remark from 'remark';
import { Node } from 'unist';

import plugin from '../';
import { CSS_CLASS } from '../constants';
import gatsbyRemarkImagesFixtures from './__fixtures__/gatsby-remark-images.json';

const warnDummy = jest.fn();

const parse = (content: string): Node => remark().parse(content);

const run = async (md: string | Node, options: Options = {}): PluginResult =>
  plugin(
    {
      markdownAST: typeof md === 'string' ? parse(md) : md,
      reporter: { warn: warnDummy }
    },
    options
  );

describe('Given the userOption styleAttributes', () => {
  it('warns about styleAttributes option being deprecated', async () => {
    const [, [image]] = await run("![remove](./remove.png '#height=10px')", {
      styleAttributes: []
    });
    expect(warnDummy).toHaveBeenCalledWith(
      '[gatsby-remark-image-attributes] The styleAttributes option is deprecated. The plugin uses all CSS properties.'
    );
    expect(image.value).toMatchSnapshot();
  });
});

describe('Given mdAST Image nodes', () => {
  it('ignores images without title', async () => {
    const [, [image]] = await run('![no title](./none.png)');
    expect(image).toBe(undefined);
  });

  it('ignores titles without key=value; pairs', async () => {
    const [, [image]] = await run(
      "![title](./title.png '#standard syntax title')"
    );
    expect(image).toBe(undefined);
  });

  it('removes attributes from the title', async () => {
    const [, [image]] = await run("![remove](./remove.png '#height=10px')");
    expect(image.title).toBe(null);
    expect(image.value).toMatchSnapshot();
  });

  it('sets title to value of a title= attribute', async () => {
    const [, [image]] = await run(
      "![fromAttribute](./attribute.png '#width=32px;title=title from attribute;color=blue')"
    );
    expect(image.title).toEqual('title from attribute');
    expect(image.value).toMatchSnapshot();
  });

  it("doesn't trip over values with spaces", async () => {
    const [, [image]] = await run(
      "![gatsby](./gatsby-logo.png '#width=10px;color=rgb(10, 10, 10)')"
    );
    expect(image.value).toMatch(
      /style="width: 10px;color: rgb\(10, 10, 10\);"/
    );
    expect(image.value).toMatchSnapshot();
  });

  it("adds plugin's CSS class to the image", async () => {
    const [, [image]] = await run(
      "![gatsby](./gatsby-logo.png '#width=10px;color=rgb(10, 10, 10)')"
    );
    expect(image.value).toMatch(
      new RegExp(`^<img.*?class="${CSS_CLASS}".*?\/>`)
    );
    expect(image.value).toMatchSnapshot();
  });
});

describe('Given mdAST HTML nodes', () => {
  it('ignores nodes without <img>', async () => {
    const [, , [image]] = await run(gatsbyRemarkImagesFixtures.notAnImage.node);
    expect(image).toBe(undefined);
  });

  describe('from gatsby-remark-images', () => {
    let titleAttributeResult: PluginResult;
    let styleAttributesResult: PluginResult;
    beforeAll(async () => {
      titleAttributeResult = await run(
        gatsbyRemarkImagesFixtures.titleAttribute.node
      );
      styleAttributesResult = await run(
        gatsbyRemarkImagesFixtures.styleAttributes.node
      );
    });

    it('ignores titles without key=value; pairs', async () => {
      const [, , [image]] = await run(
        gatsbyRemarkImagesFixtures.noAttributes.node
      );
      expect(image.title).toEqual('No attributes, here');
      expect(image.value).toMatch('title="No attributes, here"');
      expect(image.value).toMatchSnapshot();
    });

    it('removes attributes from the title', async () => {
      const [, , [image]] = styleAttributesResult;
      expect(image.title).toBe(null);
      expect(image.value).toMatch('title=""');
      expect(image.value).toMatchSnapshot();
    });

    it('sets title to value of a title= attribute', async () => {
      const [, , [image]] = titleAttributeResult;
      expect(image.title).toEqual('Title from imageAttribute');
      expect(image.value).toMatch('title="Title from imageAttribute"');
      expect(image.value).toMatchSnapshot();
    });

    it("adds plugin's CSS class to the wrapper", async () => {
      const [, , [image]] = titleAttributeResult;
      expect(image.value).toMatch(new RegExp(`^<span class="${CSS_CLASS}"`));
      expect(image.value).toMatchSnapshot();
    });

    it('applies inline-block style to inline image wrappers', async () => {
      const [, , [image]] = styleAttributesResult;
      const [, , [inlineImage]] = titleAttributeResult;
      expect(inlineImage.value).toMatch(
        /^<span.*?style="display:inline-block.*?>/
      );
      expect(image.value).toMatch(/^<span.*?style="display:block.*?>/);
      expect(image.value).toMatchSnapshot();
    });

    describe('with showCaption option', () => {
      let figureImage: PluginResult;
      beforeAll(async () => {
        [[figureImage]] = await run(gatsbyRemarkImagesFixtures.figure.node);
      });

      it('wraps with a <div>', () => {
        expect(figureImage.value).toMatch(
          new RegExp(`^<div class="${CSS_CLASS}"`)
        );
        expect(figureImage.value).toMatchSnapshot();
      });

      it('sanitizes title=""', () => {
        expect(figureImage.value).toMatch(
          /title="This has a title from image attributes"/
        );
      });

      it('sanitizes the <figcaption>', () => {
        expect(figureImage.value).toMatch(
          /<figcaption.*?>This has a title from image attributes<\/figcaption>/
        );
      });
    });
  });
});
