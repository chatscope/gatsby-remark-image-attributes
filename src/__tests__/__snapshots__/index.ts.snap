// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given mdAST HTML nodes from gatsby-remark-images adds plugin's CSS class to the wrapper 1`] = `"<span class=\\"gatsby-img-attributes\\" style=\\"display:inline-block; position: absolute;right: 0;top: 0;width: 150px;height: 150px;\\"><span class=\\"gatsby-resp-image-wrapper\\"      style=\\"position: relative; display:block; margin-left: auto; margin-right: auto; max-width: 298px; \\"    > <span    class=\\"gatsby-resp-image-background-image\\"style=\\"padding-bottom: 100%; position: relative; bottom: 0; left: 0; display:block;\\"  ></span>  <img        class=\\"gatsby-resp-image-image\\" alt=\\"fork me on github\\"title=\\"Title from imageAttribute\\"src=\\"/static/712f8e858bea9bc039da2b64fda4bece/ca501/fmog.png\\"srcset=\\"/static/712f8e858bea9bc039da2b64fda4bece/772e8/fmog.png200w,/static/712f8e858bea9bc039da2b64fda4bece/ca501/fmog.png 298w\\"sizes=\\"(max-width: 298px) 100vw, 298px\\"        style=\\"width:100%;height:100%;margin:0;vertical-align:middle;position:absolute;top:0;left:0;\\"loading=\\"lazy\\"      / >    </span></span>"`;

exports[`Given mdAST HTML nodes from gatsby-remark-images applies inline-block style to inline image wrappers 1`] = `"<span class=\\"gatsby-img-attributes\\" style=\\"display:block; box-shadow: 0px 0px 10px 2px rgb(2, 2, 2);width: 150px;\\"><span      class=\\"gatsby-resp-image-wrapper\\"      style=\\"position:relative; display: block; margin-left: auto; margin-right: auto; max-width:150px; \\"    >      <spanclass=\\"gatsby-resp-image-background-image\\"    style=\\"padding-bottom: 100%;position: relative; bottom: 0; left: 0; display: block;\\"  ></span>  <img       class=\\"gatsby-resp-image-image\\"        alt=\\"test\\" title=\\"\\"src=\\"/static/7611aa2af5f5789657c9d4c68a051876/8a4e8/gatsby-logo.png\\"srcset=\\"/static/7611aa2af5f5789657c9d4c68a051876/8a4e8/gatsby-logo.png 150w\\"        sizes=\\"(max-width: 150px) 100vw, 150px\\"        style=\\"width:100%;height:100%;margin:0;vertical-align:middle;position:absolute;top:0;left:0;\\"    loading=\\"lazy\\"      / ></span></span>"`;

exports[`Given mdAST HTML nodes from gatsby-remark-images ignores titles without key=value; pairs 1`] = `"<span class=\\"gatsby-img-attributes\\" style=\\"display:inline-block; width: 150px;\\"><span class=\\"gatsby-resp-image-wrapper\\" style=\\"position: relative; display:block; margin-left: auto; margin-right: auto; max-width: 150px; \\"> <span class=\\"gatsby-resp-image-background-image\\"style=\\"padding-bottom: 100%; position: relative; bottom: 0; left: 0; display:block;\\"></span><img class=\\"gatsby-resp-image-image\\" alt=\\"no attributes\\" title=\\"No attributes, here\\" src=\\"/static/7611aa2af5f5789657c9d4c68a051876/8a4e8/gatsby-logo.png\\"srcset=\\"/static/7611aa2af5f5789657c9d4c68a051876/8a4e8/gatsby-logo.png 150w\\" sizes=\\"(max-width: 150px) 100vw, 150px\\" style=\\"width:100%;height:100%;margin:0;vertical-align:middle;position:absolute;top:0;left:0;\\" loading=\\"lazy\\"/ ></span></span>"`;

exports[`Given mdAST HTML nodes from gatsby-remark-images removes attributes from the title 1`] = `"<span class=\\"gatsby-img-attributes\\" style=\\"display:block; box-shadow: 0px 0px 10px 2px rgb(2, 2, 2);width: 150px;\\"><span      class=\\"gatsby-resp-image-wrapper\\"      style=\\"position:relative; display: block; margin-left: auto; margin-right: auto; max-width:150px; \\"    >      <spanclass=\\"gatsby-resp-image-background-image\\"    style=\\"padding-bottom: 100%;position: relative; bottom: 0; left: 0; display: block;\\"  ></span>  <img       class=\\"gatsby-resp-image-image\\"        alt=\\"test\\" title=\\"\\"src=\\"/static/7611aa2af5f5789657c9d4c68a051876/8a4e8/gatsby-logo.png\\"srcset=\\"/static/7611aa2af5f5789657c9d4c68a051876/8a4e8/gatsby-logo.png 150w\\"        sizes=\\"(max-width: 150px) 100vw, 150px\\"        style=\\"width:100%;height:100%;margin:0;vertical-align:middle;position:absolute;top:0;left:0;\\"    loading=\\"lazy\\"      / ></span></span>"`;

exports[`Given mdAST HTML nodes from gatsby-remark-images sets title to value of a title= attribute 1`] = `"<span class=\\"gatsby-img-attributes\\" style=\\"display:inline-block; position: absolute;right: 0;top: 0;width: 150px;height: 150px;\\"><span class=\\"gatsby-resp-image-wrapper\\"      style=\\"position: relative; display:block; margin-left: auto; margin-right: auto; max-width: 298px; \\"    > <span    class=\\"gatsby-resp-image-background-image\\"style=\\"padding-bottom: 100%; position: relative; bottom: 0; left: 0; display:block;\\"  ></span>  <img        class=\\"gatsby-resp-image-image\\" alt=\\"fork me on github\\"title=\\"Title from imageAttribute\\"src=\\"/static/712f8e858bea9bc039da2b64fda4bece/ca501/fmog.png\\"srcset=\\"/static/712f8e858bea9bc039da2b64fda4bece/772e8/fmog.png200w,/static/712f8e858bea9bc039da2b64fda4bece/ca501/fmog.png 298w\\"sizes=\\"(max-width: 298px) 100vw, 298px\\"        style=\\"width:100%;height:100%;margin:0;vertical-align:middle;position:absolute;top:0;left:0;\\"loading=\\"lazy\\"      / >    </span></span>"`;

exports[`Given mdAST HTML nodes from gatsby-remark-images with showCaption option wraps with a <div> 1`] = `
"<div class=\\"gatsby-img-attributes\\" style=\\"width: 64px;\\"><figure class=\\"gatsby-resp-image-figure\\" style=\\"\\">
    <span
class=\\"gatsby-resp-image-wrapper\\"
      style=\\"position: relative; display:block; margin-left: auto; margin-right: auto; max-width: 150px; \\"
    >
 <span
    class=\\"gatsby-resp-image-background-image\\"
style=\\"padding-bottom: 100%; position: relative; bottom: 0; left: 0; display:block;\\"
  ></span>
  <img
        class=\\"gatsby-resp-image-image\\"
 alt=\\"title from attribute\\"
        title=\\"This has a title from image attributes\\"
src=\\"/static/7611aa2af5f5789657c9d4c68a051876/8a4e8/gatsby-logo.png\\"
srcset=\\"/static/7611aa2af5f5789657c9d4c68a051876/8a4e8/gatsby-logo.png 150w\\"
        sizes=\\"(max-width: 150px) 100vw, 150px\\"
        style=\\"width:100%;height:100%;margin:0;vertical-align:middle;position:absolute;top:0;left:0;\\"
    loading=\\"lazy\\"
      / >
    </span>
    <figcaption class=\\"gatsby-resp-image-figcaption\\">This has a title from image attributes</figcaption>
  </figure></div>"
`;

exports[`Given mdAST Image nodes adds plugin's CSS class to the image 1`] = `"<img src=\\"./gatsby-logo.png\\" class=\\"gatsby-img-attributes\\" style=\\"width: 10px;color: rgb(10, 10, 10);\\" alt=\\"gatsby\\" title=\\"\\"/>"`;

exports[`Given mdAST Image nodes doesn't trip over values with spaces 1`] = `"<img src=\\"./gatsby-logo.png\\" class=\\"gatsby-img-attributes\\" style=\\"width: 10px;color: rgb(10, 10, 10);\\" alt=\\"gatsby\\" title=\\"\\"/>"`;

exports[`Given mdAST Image nodes removes attributes from the title 1`] = `"<img src=\\"./remove.png\\" class=\\"gatsby-img-attributes\\" style=\\"height: 10px;\\" alt=\\"remove\\" title=\\"\\"/>"`;

exports[`Given mdAST Image nodes sets title to value of a title= attribute 1`] = `"<img src=\\"./attribute.png\\" class=\\"gatsby-img-attributes\\" style=\\"width: 32px;color: blue;\\" alt=\\"fromAttribute\\" title=\\"title from attribute\\"/>"`;

exports[`Given the userOption styleAttributes warns about styleAttributes option being deprecated 1`] = `"<img src=\\"./remove.png\\" class=\\"gatsby-img-attributes\\" style=\\"height: 10px;\\" alt=\\"remove\\" title=\\"\\"/>"`;
